userspace/libsinsp/plugin.cpp

```diff
@@ -701,8 +711,19 @@ bool sinsp_plugin::resolve_dylib_symbols(void *handle, std::string &errstr)
 	// Others are not and the values will be checked when needed.
 	(*(void **) (&m_plugin_info.init)) = getsym(handle, "plugin_init", errstr);
 	(*(void **) (&m_plugin_info.destroy)) = getsym(handle, "plugin_destroy", errstr);
-	(*(void **) (&m_plugin_info.get_fields)) = getsym(handle, "plugin_get_fields", errstr);
-	(*(void **) (&m_plugin_info.extract_fields)) = getsym(handle, "plugin_extract_fields", errstr);
+	switch (type())
+	{
+		case TYPE_SOURCE_PLUGIN:
+		case TYPE_EXTRACTOR_PLUGIN:
+			(*(void **) (&m_plugin_info.get_fields)) = getsym(handle, "plugin_get_fields", errstr);
+			(*(void **) (&m_plugin_info.extract_fields)) = getsym(handle, "plugin_extract_fields", errstr);
+			break;
+		case TYPE_CAPTURE_PLUGIN:
+		default:
+			(*(void **) (&m_plugin_info.get_fields)) = NULL;
+			(*(void **) (&m_plugin_info.extract_fields)) = NULL;
+			break;
```

Comment:
This feels vaguely arbitrary. If I have a data source that supports both live capture and savefiles, I need to create three plugins to support both use cases (you need a separate extractor plugin to do any filtering on savefiles).

It almost feels like plugin types are an arbitrary distinction too (is there a technical reason why a single plugin can't implement all three interfaces?).

Or I'm wrong and happy to be corrected.

Reply:
I see your point here.

In source plugins this problem has been mitigated by making the `get_fields` and `extract_fields` symbols optional. However, this is different from extractor-only plugins, because source plugins are likely to extract fields only from events of their source type. For instance, I would not expect the `cloudtrail` plugin to extract fields from the `dummy` event source.

Accordingly, the rationale behind capture plugins was that they were not meant to have extraction-related features because they are not tied to any specific source type. They can potentially generate events of any kind, syscall events included. Given this semantic mismatch, we opted to leave the design simple and shape capture plugin as a simple event producer.

I personally don't have a strong opinion about this, but this was the reasoning behind this choice. I think this friction is more related to how the plugin system is designed as a whole. Plugins can have only one type (returned by the `get_type` symbol), so we can't have a real "all-in-one" plugin solution. The only way to go near that, is to add optional symbols and decide what the semantics of a certain plugin type should be.

Reply:
Capture plugins may be agnostic of the content they provide. Since they may stream more data sources simultaneously (likewise with a trace file), a plugin author may be tempted to implement several extraction logic inside a single plugin. We want to avoid that.

On the other hand, a source plugin is always bound to a single data source, so it may be convenient to implement the extraction logic in the same plugin (which is optional anyway).

In your example, you won't need three plugins. You will need just two: a capture plugin and a source plugin. You don't need a separate extractor plugin because the source plugin (implementing the extraction interface) can be used for extraction only.

Reply:
Can you use both a source plugin (for extraction only) and a capture plugin at the same time?

Reply:
Ideally, many plugins might be registered at the same time in sinsp, with only one of them selected as input. An input will either be a source or a capture plugin. You can see how that happens [here](https://github.com/falcosecurity/libs/pull/159/files#diff-633539375239ac63e4a9e3358f0852d437b22e9c9f2c96de5e04e346b0b4784fR1699).

If the input is a capture plugin, any other registered source plugin might be used to extract the fields of their source, along with the other extractor plugins loaded. If the input is a source plugin, extraction might happen either from the plugin itself or from any other source-compatible extractor plugin. Example: if a capture plugin generates events of the `aws_cloudtrail` source, then the `cloudtrail` plugin will be necessary to extract the `ct.*` fields (they will be `<NA>` if `cloudtrail` is not loaded).

The way this will be supported in practice will be more of a design choice for the consumers.

Reply:
Since the primary driver for capture plugins is to handle the very special use case of accessing remote trace files through a proxy, and we also think that the only likely implementation of a capture plugin will be the one we write, I think we should treat the capture plugin very differently than other plugins.  In this case, for example, we would never expect a capture plugin to ever resolve fields, so why even have it in the struct? Said differently, why should we try to fit the API used by capture plugins into the same common_plugin_info used by source and extraction plugins?